File: c-cpp/src/skiplists/versioned-tower/versioned-tower.c
#include "versioned-tower.h"

/** TRAVERSE **************************************************************************************/

sl_node_t* traverse(val_t val, sl_node_t** prevs, sl_node_t* head) {
    sl_node_t* prev = head;
    sl_node_t* curr = NULL;
    sl_node_t* found = NULL;

    /* traverse down the levels of the skiplist */
    for (int level = (levelmax-1); level >= 0; level--) {
        curr = prev->nexts[level];

        /* traverse at the current level */
        while (curr->val < val) {
            prev = curr;
            curr = curr->nexts[level];
        }

        /* record prevs at each level */
        prevs[level] = prev;

        /* record if val is ever found */
        if (curr->val == val) {
            found = curr;
        }
    }

    /* val not found at any level */
    return found;
}

/** INSERT ****************************************************************************************/

bool validate_insert(sl_node_t *prev, sl_node_t *tower_to_insert, int level, verlock_t* valid_ver) {
    verlock_t prev_ver = get_version(&prev->lock);

    /* if prev is being inserted or being deleted */
    if(prev->status != 1) {
        return false;
    }

    /* if prev is already pointing to tower_to_insert at the current level */
    if(prev->nexts[level]->val == tower_to_insert->val) {
        return false;
    }

    /* if prev is no longer the appropriate prev to utilize */
    if(prev->nexts[level]->val < tower_to_insert->val) {
        return false;
    }

    /* success, so return the version at this level */
    *valid_ver = prev_ver;
    return true;
}


bool try_insert_at_level(sl_node_t* prev, sl_node_t* tower_to_insert, int level) {
    verlock_t valid_ver;
    do {
        /* pre-locking validation */
        if (!validate_insert(prev, tower_to_insert, level, &valid_ver)) {
            return false;
        }

    /* try-lock prev */
    } while (!try_lock_at_version(&prev->lock, valid_ver));

    /* set tower_to_insert's next */
    tower_to_insert->nexts[level] = prev->nexts[level];

    /* update prev's next */
    prev->nexts[level] = tower_to_insert; /* linearization point when level = 0 */

    unlock_and_increment_version(&prev->lock);
    return true;
}

int sl_insert(sl_intset_t *set, val_t val) {
    sl_node_t** prevs = pthread_getspecific(prevs_key);
    sl_node_t* tower_to_insert = NULL;
    sl_node_t* found_tower = NULL;

retry_from_traverse:
    /* traverse the skiplist */
    found_tower = traverse(val, prevs, set->head);

    /* the value already exists in the skiplist */
    if (found_tower != NULL) {
        /* wait-free traversed onto node just before being fully unlinked / deleted */
        if (found_tower->status == 2) {
            /* can not return false as that equals successful contains */
            goto retry_from_traverse;
        }
        /* insert linearizes at the start thus allows concurrent inserts */
        /* remove linearizes at the end thus allows concurrent inserts */
        return false;
    }

    /* create tower to insert */
    if (tower_to_insert == NULL) {
        int height = get_rand_level();
        ptst_t *ptst = ptst_critical_enter();
        tower_to_insert = sl_new_simple_node(val, height, ptst);
        ptst_critical_exit(ptst);

        /* start tower as locked */
        spinlock(&tower_to_insert->lock); /* guaranteed to work on new node */
    }

    /* insert at each level */
    int level = 0;
    while (level < tower_to_insert->height) {
        if (try_insert_at_level(prevs[level], tower_to_insert, level)) {
            /* insert has succeeded at this level */
            level++;

        /* re-traverse and try again with updated prevs */
        } else {
            /* race: aborting at level zero might mean another thread inserted first */
            if (level == 0){
                goto retry_from_traverse;
            }
            /* re-traverse to update list of prevs */
            traverse(val, prevs, set->head);
        }
    }

    /* mark as valid */
    tower_to_insert->status = 1;

    /* unlock tower */
    unlock_and_increment_version(&tower_to_insert->lock);

    return true;
}

/** REMOVE ****************************************************************************************/

bool validate_remove(sl_node_t *prev, sl_node_t *tower_to_remove, int level, verlock_t* valid_ver) {
    verlock_t prev_ver = get_version(&prev->lock);

    /* if prev is being inserted or being deleted */
    if(prev->status != 1) {
        return false;
    }

    /* if prev is not pointing to the tower at the current level */
    if(prev->nexts[level]->val != tower_to_remove->val) {
        return false;
    }

    /* success, so return the version at this level */
    *valid_ver = prev_ver;
    return true;
}


bool try_remove_at_level(sl_node_t* prev, sl_node_t* tower_to_remove, int level) {
    verlock_t valid_ver;
    do {
        /* pre-locking validation */
        if (!validate_remove(prev, tower_to_remove, level, &valid_ver)) {
            return false;
        }

    /* try-lock prev */
    } while (!try_lock_at_version(&prev->lock, valid_ver));

    /* update prev to skip over tower_to_remove*/
    prev->nexts[level] = tower_to_remove->nexts[level];  /* linearization point when level = 0 */

    unlock_and_increment_version(&prev->lock);
    return true;
}


int sl_remove(sl_intset_t *set, val_t val) {
    sl_node_t **prevs = pthread_getspecific(prevs_key);
    sl_node_t* found_tower = NULL;

retry_from_traverse:
    /* traverse the skiplist */
    found_tower = traverse(val, prevs, set->head);

    /* val is not found, so remove fails */
    if (found_tower == NULL) {
        return false;
    }

    /* pre-lock validation */
    verlock_t valid_ver = get_version(&found_tower->lock);

    /* found tower is already being removed, or is not fully inserted */
    if (found_tower->status != 1) {
        /* insert linearizes at the start thus blocks concurrent removes */
        /* remove linearizes at the end thus blocks concurrent removes */
        goto retry_from_traverse;
    }

    /* try-lock tower */
    if (!try_lock_at_version(&found_tower->lock, valid_ver)) {
        goto retry_from_traverse;
    }

    /* remove at each level */
    int level = (found_tower->height -1);
    while (level >= 0) {
        if (try_remove_at_level(prevs[level], found_tower, level)) {
            /* remove has succeeded at this level */
            level--;

        /* re-traverse and try again with updated prevs */
        } else {
            /* re-traverse to update list of prevs */
            traverse(val, prevs, set->head);
        }
    }

    /* mark tower as being deleted */
    found_tower->status = 2;

    /* no need to unlock tower as it is fully unlinked */
    //unlock_and_increment_version(&found_tower->lock);

    ptst_t *ptst = ptst_critical_enter();
    sl_delete_node(found_tower, ptst);
    ptst_critical_exit(ptst);

    return true;
}

/** CONTAINS **************************************************************************************/

int sl_contains(sl_intset_t *set, val_t val) {
    sl_node_t* prev = set->head;
    sl_node_t* curr;

    /* traverse down the levels of the skiplist */
    for (int level = (levelmax-1); level >= 0; level--) {
        curr = prev->nexts[level];

        /* traverse at the current level */
        while (curr->val < val) {
            prev = curr;
            curr = curr->nexts[level];
        }

        /* if val is found at the current level */
        if (curr->val == val) {
            /* success depends on if the tower is not deleted */
            return (curr->status != 2);
        }
    }

    /* val not found at any level */
    return false;
}

File: java/src/skiplists/versioned/VersionedTower.java
package skiplists.versioned;

import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;

import contention.abstractions.AbstractCompositionalIntSet;

// TODO: should inserts be linearized at start or finish? then either has to block concurrent inserts or deletes

public class VersionedTower extends AbstractCompositionalIntSet {

	private final Tower head;
	private final Tower tail;

	private static final int MAX_HEIGHT = 22; // TODO: use variable max height
												// selection
	private static final int TOP = MAX_HEIGHT - 1;

	private static final int OK = 0;
	private static final int ABORT = 1;
	
	//private volatile Integer currentHeight = 3;
	
	//static final AtomicReferenceFieldUpdater<VersionedTower, Integer> currentHeightUpdater = AtomicReferenceFieldUpdater
	//		.newUpdater(VersionedTower.class, Integer.class, "currentHeight");
	
	//private boolean updateHeight(int prevHeight, int newHeight) {
	//	return currentHeightUpdater.compareAndSet(this, prevHeight, newHeight);
	//}

	public VersionedTower() {

		tail = new Tower(Integer.MAX_VALUE, MAX_HEIGHT, MAX_HEIGHT);
		head = new Tower(Integer.MIN_VALUE, MAX_HEIGHT, MAX_HEIGHT);

		for (int i = 0; i < MAX_HEIGHT; i++) {
			head.nexts.set(i, tail);
		}

		head.height = MAX_HEIGHT;
		tail.height = MAX_HEIGHT;
		head.status = 1;
		tail.status = 1;
	}

	/*
	 * TRAVERSE
	 * *****************************************************************
	 * *********************
	 */

	private Tower traverse(int from, int val, Tower[] prevs) {
		Tower prev = head;
		Tower curr = null;
		Tower found = null;

		/* traverse down the levels of the skiplist */
		for (int level = from; level >= 0; level--) {
			curr = prev.nexts.get(level);

			/* traverse at the current level */
			while (curr.val < val) {
				prev = curr;
				curr = curr.nexts.get(level);
			}

			/* record prevs at each level */
			prevs[level] = prev;

			/* record if val is ever found */
			if (curr.val == val) {
				found = curr;
			}
		}

		/* val not found at any level */
		return found;
	}

	/*
	 * INSERT
	 * *******************************************************************
	 * *********************
	 */

	private int validateInsert(Tower prev, Tower towerToInsert, int level,
			int[] state) {
		int prevVer = prev.getVersion();

		/* if prev is being inserted or being deleted */
		if (prev.status != 1) {
			state[0] = ABORT;
			return 0;
		}

		/* if prev is already pointing to towerToInsert at the current level */
		if (prev.nexts.get(level).val == towerToInsert.val) {
			state[0] = ABORT;
			return 0;
		}

		/* if prev is no longer the appropriate prev to utilize */
		if (prev.nexts.get(level).val < towerToInsert.val) {
			state[0] = ABORT;
			return 0;
		}

		/* success, so return the version at this level */
		state[0] = OK;
		return prevVer;
	}

	private boolean tryInsertAtLevel(Tower prev, Tower towerToInsert, int level) {
		int validVer;
		int[] state = { OK };
		do {
			/* pre-locking validation */
			validVer = validateInsert(prev, towerToInsert, level, state);
			if (state[0] == ABORT) {
				return false;
			}

			/* try-lock prev */
		} while (!prev.tryLockAtVersion(validVer));

		/* set towerToInsert's next */
		towerToInsert.nexts.set(level, prev.nexts.get(level));

		/* update prev's next */
		prev.nexts.set(level, towerToInsert); /*
											 * linearization point when level =
											 * 0
											 */

		prev.unlockAndIncrementVersion();
		return true;
	}

	@Override
	public boolean addInt(int val) {
		Tower[] prevs = new Tower[MAX_HEIGHT];
		Tower towerToInsert = null;
		Tower foundTower = null;
		//int height = getRandomHeight();
		//int traverseHeight = Math.max(height, currentHeight);

		retryFromTraverse: while (true) {
			towerToInsert = null;
			foundTower = null;			
			/* traverse the skiplist */
			foundTower = traverse(TOP, val, prevs);

			/* the value already exists in the skiplist */
			if (foundTower != null) {
				/*
				 * wait-free traversed onto node just before being fully
				 * unlinked / deleted
				 */
				if (foundTower.status == 2) {
					/* can not return false as that equals successful contains */
					continue retryFromTraverse;
				}
				/*
				 * insert is linearized at the start thus allows concurrent
				 * inserts
				 */
				/*
				 * remove is linearized at the end thus allows concurrent
				 * inserted
				 */
				return false;
			}

			/* create tower to insert */
			if (towerToInsert == null) {
				int height = getRandomHeight();
				towerToInsert = new Tower(val, MAX_HEIGHT, height);
				//while(height > currentHeight) {
				//	this.updateHeight(currentHeight, height);
				//}
				/* start tower as locked */
				towerToInsert.spinlock(); /* guaranteed to work on new node */
			}

			/* insert at each level */
			int level = 0;
			while (level < towerToInsert.height) {
				if (tryInsertAtLevel(prevs[level], towerToInsert, level)) {
					/* insert has succeeded at this level */
					level++;
					
					/* re-traverse and try again with updated prevs */
				} else {
					/*
					 * race: aborting at level zero might mean another thread
					 * inserted first
					 */
					if (level == 0) {
						continue retryFromTraverse;
					}
					/* re-traverse to update list of prevs */
					traverse(TOP, val, prevs);
				}
			}

			break retryFromTraverse;
		}

		/* mark as valid */
		towerToInsert.status = 1;

		/* unlock tower */
		towerToInsert.unlockAndIncrementVersion();

		return true;
	}

	/*
	 * REMOVE
	 * *******************************************************************
	 * *********************
	 */

	private int validateRemove(Tower prev, Tower towerToRemove, int level,
			int[] state) {
		int prevVer = prev.getVersion();

		/* if prev is being inserted or being deleted */
		if (prev.status != 1) {
			state[0] = ABORT;
			return 0;
		}

		/* if prev is not pointing to the tower at the current level */
		if (prev.nexts.get(level).val != towerToRemove.val) {
			state[0] = ABORT;
			return 0;
		}

		/* success, so return the version at this level */
		state[0] = OK;
		return prevVer;

	}

	private boolean tryRemoveAtLevel(Tower prev, Tower towerToRemove, int level) {
		int validVer;
		int[] state = { OK };
		do {
			/* pre-locking validation */
			validVer = validateRemove(prev, towerToRemove, level, state);
			if (state[0] == ABORT) {
				return false;
			}

			/* try-lock prev */
		} while (!prev.tryLockAtVersion(validVer));

		/* update prev to skip over towerToRemove */
		prev.nexts.set(level, towerToRemove.nexts.get(level)); /*
																 * linearization
																 * point when
																 * level = 0
																 */

		prev.unlockAndIncrementVersion();
		return true;
	}

	@Override
	public boolean removeInt(int val) {
		Tower[] prevs = new Tower[MAX_HEIGHT];
		Tower foundTower = null;

		retryFromTraverse: while (true) {
			/* traverse the skiplist */
			foundTower = traverse(TOP, val, prevs);

			/* val is not found, so remove fails */
			if (foundTower == null) {
				return false;
			}

			/* pre-lock validation */
			int validVer = foundTower.getVersion();

			/* found tower is already being removed, or is not fully inserted */
			int status;
			status = foundTower.status;
			if (status == 2) {
				/* insert linearizes at the start thus blocks concurrent removes */
				/* remove linearizes at the end thus blocks concurrent removes */
				return false;
			} else if (status == 0) {
				continue retryFromTraverse;
			}

			/* try-lock tower */
			if (!foundTower.tryLockAtVersion(validVer)) {
				continue retryFromTraverse;
			}

			break retryFromTraverse;
		}

		/* remove at each level */
		int level = (foundTower.height - 1);
		while (level >= 0) {
			if (tryRemoveAtLevel(prevs[level], foundTower, level)) {
				/* remove has succeeded at this level */
				level--;

				/* re-traverse and try again with updated prevs */
			} else {
				/* re-traverse to update list of prevs */
				traverse(TOP, val, prevs);
			}
		}

		/* mark tower as being deleted */
		foundTower.status = 2;

		/* no need to unlock tower as it is fully unlinked */
		// foundTower.unlockAndIncrementVersion();

		return true;

	}

	/*
	 * CONTAINS
	 * *****************************************************************
	 * *********************
	 */

	@Override
	public boolean containsInt(int val) {
		Tower prev = head;
		Tower curr = null;

		/* traverse down the levels of the skiplist */
		for (int level = TOP; level >= 0; level--) {
			curr = prev.nexts.get(level);

			/* traverse at the current level */
			while (curr.val < val) {
				prev = curr;
				curr = curr.nexts.get(level);
			}

			/* if val is found at the current level */
			if (curr.val == val) {
				/* success depends on if the tower is not deleted */
				/* TODO this is wrong, if it is 0 should also return false */
				return (curr.status != 2);
			}
		}

		/* val not found at any level */
		return false;
	}

	/*
	 * UTILITY
	 * ******************************************************************
	 * *********************
	 */

	@Override
	public int size() {
		int size = 0;
		Tower curr = head.nexts.get(0);
		while (curr != tail) {
			if (curr.height != 0) {
				size++;
			}
			curr = curr.nexts.get(0);
		}
		return size;
	}

	@Override
	public void clear() {
		for (int i = 0; i < MAX_HEIGHT; i++) {
			head.nexts.set(i, tail);
		}
		head.resetLocks();
		tail.resetLocks();
		assert head.height == MAX_HEIGHT;
		assert tail.height == MAX_HEIGHT;
	}

	private int getRandomHeight() {
		/* height from 1 to TOP */
		return Math
				.min(TOP, (skiplists.RandomLevelGenerator.randomLevel() + 1));
	}

	void print() {
		System.out.println("skiplist:");
		for (int i = TOP - 1; i >= 0; i--) {
			Tower next = head;
			while (next != tail) {
				System.out.print(next.val + "-- ");
				next = next.nexts.get(i);
			}
			System.out.println();
		}
		System.out.println();
	}

}
